multi.c

```diff
@@ -0,0 +1,136 @@
+#include "multi.h"
+
+#include "brute.h"
+#include "queue.h"
+#include "single.h"
+#include <assert.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+void *
+mt_password_check (void *context)
+{
+  mt_context_t *mt_context = (mt_context_t *)context;
+  task_t task;
+  st_context_t st_context = {
+    .hash = mt_context->config->hash,
+    .data = { .initialized = 0 },
+  };
+
+  while (true)
+    {
+      if (queue_pop (&mt_context->queue, &task) == S_FAILURE)
+        {
+          return NULL;
+        }
+      if (st_password_check (&task, &st_context))
+        {
+          memcpy (mt_context->password, task.password, sizeof (task.password));
+        }
+
+      if (pthread_mutex_lock (&mt_context->mutex) != 0)
```

Comment:
In current implementation we could get to a situation that mutex got locked and pthread_cond_signal failed. In this case application will halt on next call to a pthread_mutex_lock. To avoid this, you need to put unlock into cleanup handler.
Read pthread_cleanup_push for details.

It should read as:
lock ();
cleanup_push (unlock, &mutex);
code_that_could_exit;
cleanup_pop (!0);